rfcs/0191-worker-manager-launch-configs.md

```diff
@@ -0,0 +1,122 @@
+* total number of failed worker provisioning attempts / registrations
+* has any worker claimed a task
+* fail-to-success ratio over a specific time period (i.e. last hour)
+* number of non-stopped workers currently running
+
+Weight will be used to determine the likelihood of selecting a specific launch configuration for provisioning.
+
+This will be calculated at the provisioning time.
+
+### Making worker-manager extensible
+
+Worker-manager will publish additional events to Pulse to allow external systems to react:
+
+* `launch-configuration-created`
+* `launch-configuration-archived`
+* `launch-configuration-paused`
+* `launch-configuration-resumed`
+* `worker-registered`
+* `worker-failure` (provisioning or starting up failure, will include `workerPoolId` and `launchConfigurationId`)
+
+New API endpoints will be introduced:
+
+* `workerManager.getLaunchConfigs(workerPoolId)` - to retrieve all launch configurations with their statuses
+* `workerManager.getLaunchConfig(workerPoolId, launchConfigId)` - to retrieve specific launch configuration
+* `workerManager.pauseLaunchConfig(workerPoolId, launchConfigId)` - to deactivate specific active launch configuration (time bound)
+* `workerManager.resumeLaunchConfig(workerPoolId, launchConfigId)` - to resume specific active launch configuration
+
+Last two endpoints would allow external systems to pause/resume specific launch configurations based on their own criteria.
+This might be useful when dynamic adjustment of weights is not enough to prevent provisioning workers in a specific region.
+
+Existing endpoints will continue to accept the same payload as before for backward compatibility.
+
+### Expiry of launch configurations
+
+During worker pool configuration updates, previous launch configurations would be marked as `archived` and kept in the database for a certain amount of time.
+Usually they should be kept for as long as the workers, that were created with this configuration.
+Once such workers are expired and removed from db, we can remove the launch configuration as well.
+
+## Examples of weight adjusting
+
+### Scenario 1 - no failures
+
+No workers have been provisioned yet, and we have two launch configurations A and B.
+Both of them would have the same weight - `1`, so the likelihood of selecting one of them would be `50%`.
+
+After some time, there could be 10 workers running for config A, and 5 workers running for config B.
+With this information, the weight would be adjusted to `0.33` for config A and `0.66` for config B.
+
+### Scenario 2 - failures in some regions
+
+There are three launch configurations A, B and C.
+At some point, provisioning workers in region A fails with quota exceeded errors.
+Weight of A would be adjusted proportionally to the error rate - `1 - (failed / total)`.
+
+Note: To avoid permanently disabling launch config, we would only adjust the weight for a specific time period (i.e. *last hour*).
```

Comment:
This makes sense, too, as long as that ramp-up is gradual. I can imagine a case where region A goes down and tasks start piling up. One hour later, the weight adjustment disappears and suddenly worker-manager tries to create 100 workers in region A, all of which fail. Better to start with just a few and "see how it goes".

Reply:
Thanks Dustin. 
I think we should separate those two problems - adjusting weight between regions and slowing down the provision rate

I tend to update this RFC to make it _only_ about the weight distribution, and leave the scaling ratio logic for an upcoming one.

Unless you have something in mind? 
If for an updated worker pool, all launch configs would have a lower weight, they would still be equally-likely selected for provisioning. 

What if ... we turn this weight to be dual-purpose .. give the likelihood and scaling ratio 🤔 
so for any weight < 1, it will act as a `scalingRatio` also .. nah, probably too complicated :)

Reply:
I think it's worth keeping things simple. And, scaling ratio is distinct enough to leave out of this RFC.